VMap IllegalStateException: Failed to find a hash

Test data: http://35.247.76.217:8095/swirlds-automation/develop/6N/Migration/20230828-091035-GCP-Daily-Migration-6N/FCM-Migration-NewNodesFromSavedState-2.5k-10m/

```
2023-08-28 09:25:42.999 168      INFO  STATE_TO_DISK    <<platform-core: signed-state-file-manager 5>> SignedStateFileWriter: Finished writing state for round 17296 to disk. Reason: PERIODIC_SNAPSHOT, directory: /opt/hgcapp/services-hedera/HapiApp2.0/data/saved/com.swirlds.demo.migration.MigrationTestingToolMain/5/123/17296 {"round":17296,"freezeState":false,"reason":"PERIODIC_SNAPSHOT","directory":"file:///opt/hgcapp/services-hedera/HapiApp2.0/data/saved/com.swirlds.demo.migration.MigrationTestingToolMain/5/123/17296/"} [com.swirlds.logging.payloads.StateSavedToDiskPayload]
2023-08-28 09:25:57.781 316      FATAL EXCEPTION        <<platform-core: state-hash-sign 5>> ManualWiring: Fatal error, node will shut down. Reason: Exception occurred during SignedState hashing {} [com.swirlds.logging.payloads.FatalErrorPayload]
   java.base/java.lang.Thread.getStackTrace(Thread.java:1610)
   com.swirlds.platform.components.wiring.ManualWiring.logFatalError(ManualWiring.java:235)
   com.swirlds.platform.components.wiring.ManualWiring.handleFatalError(ManualWiring.java:213)
   com.swirlds.platform.state.signed.SignedStateHasher.hashState(SignedStateHasher.java:95)
   com.swirlds.platform.components.state.DefaultStateManagementComponent.newSignedStateFromTransactions(DefaultStateManagementComponent.java:381)
   com.swirlds.common.threading.framework.internal.QueueThreadImpl.doWork(QueueThreadImpl.java:253)
   com.swirlds.common.threading.framework.internal.StoppableThreadImpl.doWork(StoppableThreadImpl.java:614)
   com.swirlds.common.threading.framework.internal.StoppableThreadImpl.run(StoppableThreadImpl.java:215)
   java.base/java.lang.Thread.run(Thread.java:833)

java.util.concurrent.ExecutionException: java.lang.RuntimeException: exception encountered while hashing virtual tree
	at com.swirlds.common.merkle.hash.FutureMerkleHash.rethrowException(FutureMerkleHash.java:83) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.merkle.hash.FutureMerkleHash.get(FutureMerkleHash.java:93) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.merkle.hash.FutureMerkleHash.get(FutureMerkleHash.java:30) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.platform.state.signed.SignedStateHasher.hashState(SignedStateHasher.java:84) ~[swirlds-platform-core-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.platform.components.state.DefaultStateManagementComponent.newSignedStateFromTransactions(DefaultStateManagementComponent.java:381) ~[swirlds-platform-core-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.threading.framework.internal.QueueThreadImpl.doWork(QueueThreadImpl.java:253) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.threading.framework.internal.StoppableThreadImpl.doWork(StoppableThreadImpl.java:614) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.threading.framework.internal.StoppableThreadImpl.run(StoppableThreadImpl.java:215) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.lang.RuntimeException: exception encountered while hashing virtual tree
	at com.swirlds.virtualmap.internal.hash.VirtualHasher.hashSubTree(VirtualHasher.java:708) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.virtualmap.internal.hash.VirtualHasher.hash(VirtualHasher.java:424) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.virtualmap.internal.merkle.VirtualRootNode.computeHash(VirtualRootNode.java:1264) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.hashCopy(VirtualPipeline.java:388) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.virtualmap.internal.merkle.VirtualRootNode.getHash(VirtualRootNode.java:1204) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.merkle.hash.MerkleHashBuilder.hashSubtree(MerkleHashBuilder.java:206) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.merkle.hash.MerkleHashBuilder.lambda$createHashingRunnable$1(MerkleHashBuilder.java:173) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	... 1 more
Caused by: java.lang.IllegalStateException: Failed to find a hash for 66548, which is a sibling of 66547
	at com.swirlds.virtualmap.internal.hash.VirtualHasher.lambda$hashSubTree$2(VirtualHasher.java:667) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
```

--- a/platform-sdk/platform-apps/tests/MigrationTestingTool/src/main/java/com/swirlds/demo/migration/MigrationTestingToolMain.java
+++ b/platform-sdk/platform-apps/tests/MigrationTestingTool/src/main/java/com/swirlds/demo/migration/MigrationTestingToolMain.java
@@ -59,7 +59,7 @@ public class MigrationTestingToolMain implements SwirldMain {
     private double toCreate = 0;
     private long lastEventTime = System.nanoTime();
 
-    public static final long SOFTWARE_VERSION = 2L;
+    public static final long SOFTWARE_VERSION = 3L;
     public static final BasicSoftwareVersion PREVIOUS_SOFTWARE_VERSION = new BasicSoftwareVersion(SOFTWARE_VERSION - 1);
     private final BasicSoftwareVersion softwareVersion = new BasicSoftwareVersion(SOFTWARE_VERSION);
 

--- a/platform-sdk/platform-apps/tests/MigrationTestingTool/src/main/java/com/swirlds/demo/migration/MigrationTestingToolState.java
+++ b/platform-sdk/platform-apps/tests/MigrationTestingTool/src/main/java/com/swirlds/demo/migration/MigrationTestingToolState.java
@@ -40,6 +40,7 @@ import com.swirlds.demo.migration.virtual.AccountVirtualMapKeySerializer;
 import com.swirlds.demo.migration.virtual.AccountVirtualMapValue;
 import com.swirlds.demo.migration.virtual.AccountVirtualMapValueSerializer;
 import com.swirlds.merkle.map.MerkleMap;
+import com.swirlds.merkledb.MerkleDb;
 import com.swirlds.merkledb.MerkleDbDataSourceBuilder;
 import com.swirlds.merkledb.MerkleDbTableConfig;
 import com.swirlds.virtualmap.VirtualMap;
@@ -209,6 +210,9 @@ public class MigrationTestingToolState extends PartialNaryMerkleInternal impleme
                 new AccountVirtualMapKeySerializer(),
                 (short) 1,
                 new AccountVirtualMapValueSerializer());
+        // to make it work for the multiple node in one JVM case, we need reset the default instance path every time
+        // we create another instance of MerkleDB.
+        MerkleDb.resetDefaultInstancePath();
         final VirtualDataSourceBuilder<AccountVirtualMapKey, AccountVirtualMapValue> dsBuilder =
                 new MerkleDbDataSourceBuilder<>(tableConfig);
         setVirtualMap(new VirtualMap<>("virtualMap", dsBuilder));

--- a/platform-sdk/swirlds-common/src/main/java/com/swirlds/common/io/utility/RecycleBinImpl.java
+++ b/platform-sdk/swirlds-common/src/main/java/com/swirlds/common/io/utility/RecycleBinImpl.java
@@ -132,7 +132,8 @@ public class RecycleBinImpl implements RecycleBin, Startable, Stoppable {
     @Override
     public void recycle(@NonNull final Path path) throws IOException {
         if (!Files.exists(path)) {
-            logger.warn(EXCEPTION.getMarker(), "Cannot recycle non-existent file: {}", path);
+            // FUTURE WORK: https://github.com/hashgraph/hedera-services/issues/8621
+            logger.warn(STARTUP.getMarker(), "Cannot recycle non-existent file: {}", path);
             return;
         }
 

